working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1803,21 +1811,44 @@ type.**
 
 ### Exhaustiveness and reachability
 
-A switch is *exhaustive* if all possible values of the matched value's type will
-definitely match at least one case, or there is a default case. Dart currently
-shows a warning if a switch statement on an enum type does not have cases for
-all enum values (or a default). This is helpful for code maintainance: when you
-add a new value to an enum type, the language shows you every switch statement
-that may need a new case to handle it.
+A switch is *exhaustive* if all possible values of the matched value's static
+type will definitely match at least one case, or there is a default case. Dart
+currently shows a warning if a switch statement on an enum type does not have
+cases for all enum values (or a default). This is helpful for code maintainance:
+when you add a new value to an enum type, the language shows you every switch
+statement that may need a new case to handle it.
 
 This checking is even more important with this proposal. Exhaustiveness checking
 is a key part of maintaining code written in an algebraic datatype style. It's
 the functional equivalent of the error reported when a concrete class fails to
 implement an abstract method.
 
 Exhaustiveness checking over arbitrarily deeply nested record and extractor
-patterns can be complex, so the proposal for that is in a [separate
-document][exhaustiveness].
+patterns is complex, so the proposal to define how it works is in a [separate
+document][exhaustiveness]. That tells us if the cases in a switch statement
+or expression are exhaustive or not. Given that:
+
+*   It is a compile-time error if the cases in a switch expression are not
+    exhaustive. *Since an expression must yield a value, the only other option
+    is to throw an error and most Dart users prefer to catch those kinds of
+    mistakes at compile time.*
+
+*   It is a compile-time error if the static type of the matched value in a
+    switch statement is an *exhaustive type* and the cases are not exhaustive.
+    An exhaustive type is:
+
+    *   `bool`
+    *   `Null`
+    *   A type whose declaration is marked sealed
+    *   `T?` where `T` is exhaustive
+    *   `FutureOr<T>` for some type `T` that is exhaustive
```

Comment:
I admit I haven't put a *lot* of thought into this case, but it felt like the right behavior to me.

> Would a general rule be that a union is exhaustive if any part of it is exhaustive?
That could be argued: This type must be exhausted. This type is part of the possible values you switch over. Ergo, this switch must be exhaustive.

Yes, I think so. One way to think of it is, when you mark a type exhaustive, you're saying, "I *want* you to yell at me whenever this type is not used exhaustively in a switch." Since you have opted *in* to that compile-time checking, the compiler should persist that preference as well as its able. That suggests that whenever it shows up in a union type, it taints the entire type. Otherwise, the alternative would be to silently drop the exhaustiveness checks on the floor, which is unlikely to be what the user wants.

> If so, could we say that a switch must "exhaust the exhaustive type", without needing to be exhaustive in general?

I think simplest is best and just say it must be exhaustive. Otherwise, you could promote but still not actually use the thing.

See also: https://github.com/dart-lang/language/issues/2179